api-module-library/sharepoint/manager.js

```diff
@@ -0,0 +1,186 @@
+            instance.entity = await Entity.findById(params.entityId);
+            const credential = await Credential.findById(
+                instance.entity.credential
+            );
+            instance.credential = credential;
+            sharepointParams.access_token = credential.accessToken;
+            sharepointParams.refresh_token = credential.refreshToken;
+        }
+        instance.api = new Api(sharepointParams);
+
+        return instance;
+    }
+
+    async testAuth() {
+        let validAuth = false;
+        try {
+            if (await this.api.listSites()) validAuth = true;
+        } catch (e) {
+            flushDebugLog(e);
+        }
+        return validAuth;
+    }
+
+    getAuthorizationRequirements() {
+        return {
+            url: this.api.getAuthUri(),
+            type: 'oauth2',
+        };
+    }
+
+    async processAuthorizationCallback(params) {
+        const code = get(params.data, 'code', 'test');
+        await this.api.getTokenFromCode(code);
+        const authCheck = await this.testAuth();
+        if (!authCheck) throw new Error('Authentication failed');
+
+        const userDetails = await this.api.getUser();
+        // TODO determine if there's a good flag to make for this, where we have individual tokens vs. org/tenant tokens
+        //  The issue here is that the Entity should reflect "on whose behalf are we making api requests", and in the
+        //  individual user case, it's a user. In the org/tenant case, it's a tenant.
+        //  The catch is that personal microsoft users do not have an org. So the graph API throws a 500 error.
+        // const orgDetails = await this.api.getOrganization();
+
+        await this.findOrCreateEntity({
+            externalId: userDetails.id,
+            name: `${userDetails.displayName} (${userDetails.userPrincipalName})`
+        });
+        return {
+            entity_id: this.entity.id,
+            credential_id: this.credential.id,
+            type: Manager.getName(),
+        };
+    }
+
+    async findOrCreateEntity(params) {
+        const externalId = get(params, 'externalId');
+        const name = get(params, 'name');
+
+        // TODO-new... this doesn't allow for multiple entities for a specific User.
+        const search = await Entity.find({
+            user: this.userId,
+            externalId,
+        });
+        if (search.length === 0) {
+            // validate choices!!!
+            // create entity
+            const createObj = {
+                credential: this.credential.id,
+                user: this.userId,
+                name,
+                externalId,
+            };
+            this.entity = await Entity.create(createObj);
+        } else if (search.length === 1) {
+            this.entity = search[0];
+        } else {
+            const message = 'Multiple entities found with the same external ID: ' + externalId;
+            debug(message);
+            throw new Error(message);
```

Comment:
An error is being thrown now.

And replaced the `portalId` reference which came from nowhere, with `externalId`